### PR TITLE
add inherit_validation_context flag to AssociatedValidator

### PR DESCRIPTION
**Pull request by rafal-brize**
### Summary

This PR comes as a result of a research done to solve issue https://github.com/rails/rails/issues/23261.
This allows to pass a forward_validation_context flag to `AssociatedValidator` so that writing `parent.valid?(:custom_context)` will pass that `:custom_context` down to child validation.
This is consistent with what `:autosave` does as a default behavior https://github.com/rails/rails/blob/master/activerecord/lib/active_record/autosave_association.rb#L328. Doing that a default behavior for `AssociatedValidator` would be a breaking change.

### CODE REVIEW

**Comment by brunzino** [comment]
Bump. (I'd love to see this implemented) Thanks, guys, for your work!

**Comment by wenderjean** [comment]
any update about this?

File: activerecord/lib/active_record/validations/associated.rb
Comment (rafaelfranca) [inline activerecord/lib/active_record/validations/associated.rb]: Do we need this memoization here?

**Review by rafaelfranca — COMMENTED** [review]
This looks good. Can you rebase?

**Comment by rafal-brize** [comment]
@rafaelfranca Updated.

**Review by sikachu — APPROVED** [review]

**Comment by itsNikolay** [comment]
@rafal-brize could you rebase the PR again? 
I met this issue too. And I made 100% the same PR :) 
But you're already done it 👍

**Comment by itsNikolay** [comment]
@rafaelfranca could you review it? 
this bug hangs almost 2 years

**Comment by wenderjean** [comment]
@rafal-brize Do you mind to rebase this one more time? @rafaelfranca Don't you think this is a really important fix?

**Comment by pedrofurtado** [comment]
@rafaelfranca Any news about this PR? That is a simple, but SO IMPORTANT fix :+1:

**Comment by franciscomxs** [comment]
Same issue here.
Good job @rafal-brize

**Comment by Tashows** [comment]
@rafaelfranca @itsNikolay bump!

**Comment by Ancez** [comment]
Hello, any news regarding this issue?
@rafal-brize could you please rebase?

@rafaelfranca @itsNikolay bumppp

**Comment by obfuscoder** [comment]
One more rebase(TM). Hope this gets merged soon ;-)

**Comment by jesseclark** [comment]
👍 This would be nice to have. 

It was a bit surprising to find out that this wasn't the default behavior since it is the default for `autosave`.

**Comment by pedrofurtado** [comment]
Anyone can reopen this PR?

**Comment by ryantk** [comment]
I just set out using validation contexts for a feature I am working on today and was bitten by this issue.
Would love it if some superhero fixed the conflicts and merged for me :D

File: activerecord/CHANGELOG.md
Comment (sikachu) [inline activerecord/CHANGELOG.md:1]: Would you mind updating the CHANGELOG to include some more information?
I think what you have [here](https://github.com/rails/rails/pull/24135/files#diff-505967108be6686c7d9744aa378f0575R65-R66) is very clear on what this feature does.

**Review by sikachu — CHANGES_REQUESTED** [review]
I just came back to look at this ... and just noticed a few things.

- This PR needs rebase again (sorry).
- I think it would be nice if CHANGELOG contains a bit more detail about this feature.
- Now, I'm starting to question the naming of this flag a bit. It took me a while to realize what `forward` means in this context. Would `inherited_validation_context` be a better name since it's _inherited_ from the parent?

Thank you for your work on this. I'll try to get someone's attention to get this merge in once you take care of those issues.

**Comment by pedrofurtado** [comment]
Happy to see it merged asap! 🎉 🏆 🥇

**Review by sikachu — APPROVED** [review]

**Review by jeremy — CHANGES_REQUESTED** [review]
Are there cases where *don't* want to inherit the parent record's validation context? Should this default to true—or not a config option at all?

**Comment by pedrofurtado** [comment]
Rails bot, please reopen this 😄

**Comment by searls** [comment]
I just got bitten by this and would love to see some kind of solution merged (I'm currently stuck trying to think through ways to pull a bunch of validation logic out of AR models as a result, which seems suboptimal)

I agree with @jeremy's point above that this sure seems like it should be the default behavior, even if it's ultimately a breaking change I highly doubt anybody is relying on the current behavior intentionally and I'd sooner classify this as a bug.

**Comment by emolayi** [comment]
I also wanted this functionality on my `has_one` associations and have managed to get it working by using a combination of the `required` option in combination with `validates_associated`:
```
class Request < ApplicationRecord
  ...

  has_one :participant, required: true, dependent: :destroy
  accepts_nested_attributes_for :participant, allow_destroy: true
  validates_associated :participant

  with_options on: :summary do
    validates :location, :visit_type, presence: true
    ....
  end
...
end

class Participant < ApplicationRecord
...
  belongs_to :request

  with_options on: :summary do
    validates :first_name, :last_name, :dob, presence: true
...
  end
...
end

request.valid?(:summary)
# error on request.location as well as request.patient.first_name, request.patient.last_name, request.patient.dob
```

**Comment by pedrofurtado** [comment]
Commenting on this PR for bot don't close it 👍

**Comment by intrip** [comment]
@rafal-brize are you still interested in moving this forward?

**Comment by pedrofurtado** [comment]
Happy to see this merged, if possible 🎉 🍻

**Comment by rafal-brize** [comment]
Hey @intrip the PR is from 2016, since then it was reviewed by number of maintainers, with different point of view on this.
Shall it become a new default (breaking change) or an opt-in behaviour. 

I am happy to work on this again, provided there is a maintainer that would guide me through this process.

My take on it is to make it a opt-in first (like i did) and maybe a default later.

**Comment by intrip** [comment]
@rafal-brize I understand this is open for a lot of time and you took your time working into this, that's why I'd really like to un-stale this situation.

I'd start improving the changelog as @jeremy suggested https://github.com/rails/rails/pull/24135#discussion_r474895991 and fixing conflicts such that the PR can be reviewed as up-to-date.

As per opt-in or not my take on this is to make it opt-in but I'm not the decision maker so we need to wait for further feedback.

**Comment by pedrofurtado** [comment]
Can someone reopen please?👍

**Comment by mjbellantoni** [comment]
Could someone reopen this?

**Comment by erikdstock** [comment]
I realize this is now 7 years old but it just bit me too.

**Comment by superdaminou** [comment]
Hi, having an issue today and this would be really great. Do we have any news on this feature ?

**Comment by austenmadden** [comment]
I ended up finding and doing something similar [here](https://github.com/rails/rails/pull/46238) as well if anyone is interested.

**Comment by austenmadden** [comment]
@rafaelfranca @rafal-brize this PR should be ready to go per the comment here: https://github.com/rails/rails/pull/46238#issuecomment-1782110991

**Pull request closed without merging**
